Fix Indentation in several statements

jorgerangel-msft opened the pull request:
Fixes the indentation in several of the statements and also fixes writing arguments in the code writer.

azure-sdk commented:
**API change check**

API changes are not detected in this pull request.

JoshLove-msft commented on packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Statements/SwitchCaseStatement.cs:
```diff
@@ -43,8 +43,10 @@ internal override void Write(CodeWriter writer)
 
-            using var scope = writer.ScopeRaw(":", "", false);
-            Statement.Write(writer);
+            using (writer.ScopeRaw(":", string.Empty, false))
```
nit: why did we change to using a block scope here?

JoshLove-msft replied:
Is it necessary to pass false here? Since we are not at the beginning of the line I expected that we wouldn't add the indentation anyway. I ask because I'm wondering if we can do away with the addSpaces param and instead just not add spaces when the line starts with a `#`.

jorgerangel-msft replied:
hey @JoshLove-msft, thanks for pointing this out. You're right, the block isn't needed. I went ahead and reverted this back. Originally, I had made some other [changes](https://github.com/microsoft/typespec/pull/3603/files#diff-5a5e5c753fc475ae65640aacc837c2ee5c00f98684e51519a2884a1d8dda6d2e) to this statement and I pulled a snippet of that over. 

Regarding passing in `false`, that param actually refers to the `newLine` param. I went ahead and added it as a named parameter for clarity. If it's not added, then an additional new line is added at the end of each switch case scope. Here is that change outlined in the unit test I added:
![image](https://github.com/microsoft/typespec/assets/102122018/d737ce6a-0ad5-4435-9dd3-208785184e5a)

I also went ahead and took your tip of just checking if the span character is not equal to `#`, rather than an additional param to these APIs. Thank you.

JoshLove-msft commented:
Curious - how did we know the indentation was wrong? Is there something in the Test project that needs to be regened?

ArcturusZhang commented:
> Curious - how did we know the indentation was wrong? Is there something in the Test project that needs to be regened?

I believe this change is extracted from some upcoming changes, therefore the part it is intended to change does not exist now.

jorgerangel-msft commented:
> > Curious - how did we know the indentation was wrong? Is there something in the Test project that needs to be regened?
> 
> I believe this change is extracted from some upcoming changes, therefore the part it is intended to change does not exist now. But I think we could add a test case to validate the indentations

That's correct. This PR is decoupled from https://github.com/microsoft/typespec/pull/3603, which is where I discovered all these scenarios as I added support for generating serialization method bodies. I added new tests for all of these statements I added. If there's more scenarios we want to test, please do let me know.

m-nash commented on packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Writers/CodeWriter.cs:
```diff
@@ -618,8 +618,11 @@ private CodeWriter AppendRaw(ReadOnlySpan<char> span)
                 return this;
-
-            AddSpaces();
+            // pre-processor directives do not need indentation
```
Can we move this into the AddSpaces function to encapsulate all indentation logic in that function?

m-nash commented on packages/http-client-csharp/generator/Microsoft.Generator.CSharp/test/Expressions/ExpressionsTest.cs:
```diff
@@ -0,0 +1,86 @@
+    internal class ExpressionsTest
+    {
+        private readonly string _mocksFolder = "Mocks";
+        private readonly string _licenseString = "// License string";
```
You can just pass `false` to your ToString call and avoid always needing to compare this header portion

m-nash commented on packages/http-client-csharp/generator/Microsoft.Generator.CSharp/test/TestData/Statements/SwitchCaseStatement.txt:
```diff
@@ -0,0 +1,17 @@
+// License string
```
rename filetype to *.cs and put these into a TestData folder then add a Compile remove tag in the csproj for the test project

m-nash commented on packages/http-client-csharp/generator/Microsoft.Generator.CSharp/test/TestData/Statements/SwitchCaseStatementWithUsing.txt:
```diff
@@ -0,0 +1,22 @@
+    switch (foo)
+    {
+        case true:
+            using (var x = new global::System.IO.MemoryStream())
```
should not be seeing MemoryStream here, you might need to add metatdata reference for typeof(object)

jorgerangel-msft replied:
That happen during the Roslyn processing right? These unit tests just test writing the statements, so there is no workspace setup or file generation called. I manually wrote these test data files, and asserted that the statements matched what I had written.

m-nash replied:
ah, thats true make sense now.

m-nash commented on packages/http-client-csharp/generator/Microsoft.Generator.CSharp/src/Writers/CodeWriter.cs:
```diff
@@ -936,19 +939,18 @@ public void WriteArguments(IEnumerable<ValueExpression> arguments, bool useSingl
             else
             {
-                WriteRawLine("(");
+                AppendRaw("(");
```
Can we add a test to validate the multiline arg case here?  I think you need more than 6 args to get into this else or explicitly set usesingleline false
